server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -189,7 +191,14 @@ public static <Response extends ReplicationResponse & WriteResponse> ActionListe
     }
 
     @Override
-    protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<BulkResponse> listener) {
+    protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<BulkResponse> outerListener) {
+        // As a work-around to support `?refresh`, explicitly replace the refresh policy with a call to the Refresh API.
+        // TODO: Replace with a less hacky approach.
+        ActionListener<BulkResponse> listener = outerListener;
+        if (DiscoveryNode.isStateless(clusterService.getSettings()) && bulkRequest.getRefreshPolicy() != WriteRequest.RefreshPolicy.NONE) {
+            listener = outerListener.delegateFailure((l, r) -> { client.admin().indices().prepareRefresh().execute(l.map(ignored -> r)); });
+            bulkRequest.setRefreshPolicy(WriteRequest.RefreshPolicy.NONE);
```

Comment:
I think for the `WAIT_UNTIL` to work properly, we'd need to also integrate the refresh listeners mechanism in Stateless, since current they don't get called back. Since we're going with work-arounds here to keep it short, I've just replaced the policy once we know we're calling a refresh afterwards anyway.